kura/org.eclipse.kura.wire.helper.provider/src/main/java/org/eclipse/kura/internal/wire/helper/WireSupportImpl.java

```diff
@@ -41,80 +50,176 @@
 
     private static final WireMessages message = LocalizationAdapter.adapt(WireMessages.class);
 
-    private final EventAdmin eventAdmin;
-
-    private List<Wire> incomingWires;
-
-    private List<Wire> outgoingWires;
-
+    private Set<WireConfiguration> wireConfigurations;
     private final WireHelperService wireHelperService;
-
-    private final WireComponent wireSupporter;
+    private final WireComponent wireComponent;
+    private final EventAdmin eventAdmin;
+    private final String componentPid;
 
     /**
-     * Instantiates a new wire support implementation.
+     * Instantiates a new {@link WireSupportImpl}
      *
-     * @param wireSupporter
-     *            the wire supporter
+     * @param wireComponent
+     *            the {@link WireComponent} instance
      * @param wireHelperService
-     *            the Wire Helper service
+     *            the {@link WireHelperService} instance
      * @param eventAdmin
-     *            the Event Admin service
+     *            the {@link EventAdmin} instance
      * @throws NullPointerException
      *             if any of the provided arguments is null
      */
-    WireSupportImpl(final WireComponent wireSupporter, final WireHelperService wireHelperService,
+    WireSupportImpl(final WireComponent wireComponent, final WireHelperService wireHelperService,
             final EventAdmin eventAdmin) {
-        requireNonNull(wireSupporter, message.wireSupportedComponentNonNull());
+        requireNonNull(wireComponent, message.wireSupportedComponentNonNull());
         requireNonNull(wireHelperService, message.wireHelperServiceNonNull());
         requireNonNull(eventAdmin, message.eventAdminNonNull());
 
-        this.outgoingWires = CollectionUtil.newArrayList();
-        this.incomingWires = CollectionUtil.newArrayList();
-        this.wireSupporter = wireSupporter;
+        this.wireComponent = wireComponent;
         this.wireHelperService = wireHelperService;
         this.eventAdmin = eventAdmin;
+        this.componentPid = this.wireHelperService.getPid(this.wireComponent);
     }
 
     /** {@inheritDoc} */
     @Override
-    public synchronized void consumersConnected(final Wire[] wires) {
-        this.outgoingWires = Arrays.asList(wires);
+    public synchronized void emit(final List<WireRecord> wireRecords) {
+        requireNonNull(wireRecords, message.wireRecordsNonNull());
+        if (!(this.wireComponent instanceof WireEmitter)) {
+            return;
+        }
+        retrieveWireConfigurationsIfAbsent();
```

Comment:
There was a synchronization issue in WireAdmin callbacks. Fixed and tested it. It will work now.